ethcore/src/engines/validator_set/multi.rs

```diff
@@ -40,42 +41,73 @@ impl Multi {
 		}
 	}
 
-	fn correct_set(&self, bh: &H256) -> Option<&Box<ValidatorSet>> {
-		match self
-			.block_number
-			.read()(bh)
-			.map(|parent_block| self
-					 .sets
-					 .iter()
-					 .rev()
-					 .find(|&(block, _)| *block <= parent_block + 1)
-					 .expect("constructor validation ensures that there is at least one validator set for block 0;
-									 block 0 is less than any uint;
-									 qed")
-				) {
-			Ok((block, set)) => {
-				trace!(target: "engine", "Multi ValidatorSet retrieved for block {}.", block);
-				Some(set)
-			},
+	fn correct_set(&self, id: BlockId) -> Option<&ValidatorSet> {
+		match self.block_number.read()(id).map(|parent_block| self.correct_set_by_number(parent_block)) {
+			Ok((_, set)) => Some(set),
 			Err(e) => {
 				debug!(target: "engine", "ValidatorSet could not be recovered: {}", e);
 				None
 			},
 		}
 	}
+
+	// get correct set by block number, along with block number at which
+	// this set was activated.
+	fn correct_set_by_number(&self, parent_block: BlockNumber) -> (BlockNumber, &ValidatorSet) {
+		let (block, set) = self.sets.iter()
+			.rev()
+			.find(|&(block, _)| *block <= parent_block + 1)
+			.expect("constructor validation ensures that there is at least one validator set for block 0;
+					 block 0 is less than any uint;
+					 qed");
+
+		trace!(target: "engine", "Multi ValidatorSet retrieved for block {}.", block);
+		(*block, &**set)
+	}
 }
 
 impl ValidatorSet for Multi {
-	fn contains(&self, bh: &H256, address: &Address) -> bool {
-		self.correct_set(bh).map_or(false, |set| set.contains(bh, address))
+	fn default_caller(&self, block_id: BlockId) -> Box<Call> {
+		self.correct_set(block_id).map(|set| set.default_caller(block_id))
+			.unwrap_or(Box::new(|_, _| Err("No validator set for given ID.".into())))
+	}
+
+	fn is_epoch_end(&self, header: &Header, block: Option<&[u8]>, receipts: Option<&[::receipt::Receipt]>)
+		-> EpochChange
+	{
+		let (set_block, set) = self.correct_set_by_number(header.number());
+
+		match set.is_epoch_end(header, block, receipts) {
+			EpochChange::Yes(num, proof) => EpochChange::Yes(set_block + num, proof),
+			other => other,
+		}
+	}
+
+	fn epoch_proof(&self, header: &Header, caller: &Call) -> Result<Vec<u8>, String> {
+		self.correct_set_by_number(header.number()).1.epoch_proof(header, caller)
+	}
+
+	fn epoch_set(&self, header: &Header, proof: &[u8]) -> Result<(u64, super::SimpleList), ::error::Error> {
```

Comment:
Makes sense, though this makes `SimpleList` a special fundamental `ValidatorSet` rather than just another trait implementation. But I guess that's fine since there are no other scenarios yet.